partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt

```diff
@@ -63,32 +64,114 @@ internal class ExprCallDynamic(
-                        if (c.target != parameterType) error("Internal Error: Cast Target does not match Function Parameter")
+                    val parametersIncludeAny = lookupTypes.any { it == PartiQLValueType.ANY }
+                    // A way to simplify logic further below. If it's empty, add something and set the processing type.
+                    if (accumulator.isEmpty()) {
+                        activelyProcessingAny = parametersIncludeAny
+                        accumulator.add(lookupTypes to candidate)
+                        return@forEach
+                    }
+                    when (parametersIncludeAny) {
+                        true -> when (activelyProcessingAny) {
+                            true -> accumulator.add(lookupTypes to candidate)
+                            false -> {
+                                activelyProcessingAny = true
+                                lookupsMutable.add(Direct.of(accumulator.toList()))
+                                accumulator.clear()
+                                accumulator.add(lookupTypes to candidate)
+                            }
+                        }
+                        false -> when (activelyProcessingAny) {
+                            false -> accumulator.add(lookupTypes to candidate)
+                            true -> {
+                                activelyProcessingAny = false
+                                lookupsMutable.add(Indirect(accumulator.toList()))
+                                accumulator.clear()
+                                accumulator.add(lookupTypes to candidate)
+                            }
+                        }
+                    }
+                }
+                // Add any remaining candidates (that we didn't submit due to not ending while switching)
+                when (accumulator.isEmpty()) {
+                    true -> { /* Do nothing! */ }
+                    false -> when (activelyProcessingAny) {
+                        true -> lookupsMutable.add(Indirect(accumulator.toList()))
+                        false -> lookupsMutable.add(Direct.of(accumulator.toList()))
+                    }
+                }
+                this.lookups = lookupsMutable
+            }
+
+            override fun get(args: List<PartiQLValueType>): Candidate? {
+                return this.lookups.firstNotNullOfOrNull { it.get(args) }
+            }
+        }
+
+        /**
+         * An O(1) structure to quickly find directly matching dynamic candidates.
+         */
+        data class Direct private constructor(val directCandidates: Map<List<PartiQLValueType>, Candidate>) : CandidateIndex {
+
+            companion object {
+                internal fun of(candidates: List<Pair<List<PartiQLValueType>, Candidate>>) = Direct(candidates.toMap())
+            }
+
+            override fun get(args: List<PartiQLValueType>): Candidate? {
+                return directCandidates[args]
+            }
```

Comment:
perhaps we choose array over list? Then its on the stack rather than a reference type. Might have marginally better lookup too